### PR TITLE
getShortestPath only calls Dijkstra's Once

### PR DESCRIPTION
**Pull request by Jmeyer1292**
This PR relates to #153. It transforms the muli-source shortest path problem to a single source shortest path problem by inserting a single, "virtual" joint into the graph with zero weight connections to all of the "real" starting points.

Results in a significant improvement in planning time for points with many starting configurations.

### CODE REVIEW

**Comment by Jmeyer1292** [comment]
**Indigo Devel**

--- Angle_disc = pi/4 ---

```
Calculating edges for 21600 total joint solutions
[ INFO] [1458849477.295537286]: Descartes found path with total cost: 71.4789
[ INFO] [1458849477.295608116]: Traj size: 1200 List size: 1200
Total time (ms): 121
Trajectory size: 1200
```

--- Angle_disc = 0.1 ---

```
[ INFO] [1458849578.370009978]: Calculating edges for 151200 total joint solutions
[ INFO] [1458849588.301413726]: Descartes found path with total cost: 69.418
[ INFO] [1458849588.301553249]: Traj size: 1200 List size: 1200
Total time (ms): 10418
Trajectory size: 1200
```

--- Angle_disc = 0.05 ---

```
[ INFO] [1458849612.849249249]: Calculating edges for 302400 total joint solutions
[ INFO] [1458849666.437148514]: Descartes found path with total cost: 68.8239
[ INFO] [1458849666.437273919]: Traj size: 1200 List size: 1200
Total time (ms): 54552
Trajectory size: 1200
```

**Single-Source Branch** 
--- Angle_disc = pi/4 ---

```
[ WARN] [1458850174.173339397]: UR5 Descartes Robot Model initialized
[ INFO] [1458850174.249559589]: Calculating edges for 21600 total joint solutions
[ INFO] [1458850174.293939183]: Descartes found path with total cost: 71.4789
[ INFO] [1458850174.294018324]: Traj size: 1200 List size: 1200
Total time (ms): 120
Trajectory size: 1200
```

--- Angle_disc = 0.1 ---

```
[ INFO] [1458850200.482154000]: Calculating edges for 151200 total joint solutions
[ INFO] [1458850201.949286930]: Descartes found path with total cost: 69.418
[ INFO] [1458850201.949372620]: Traj size: 1200 List size: 1200
Total time (ms): 1954
Trajectory size: 1200
```

--- Angle_disc = 0.05 ---

```
[ INFO] [1458850236.881528331]: Calculating edges for 302400 total joint solutions
[ INFO] [1458850243.129645277]: Descartes found path with total cost: 68.8239
[ INFO] [1458850243.129735130]: Traj size: 1200 List size: 1200
Total time (ms): 7219
Trajectory size: 1200
```

**Comment by Jmeyer1292** [comment]
@shaun-edwards The above charts total planning time including graph building and search.

You'll notice at coarse discretizations, the search time is so small compared to graph building that no improvement is seen. For large graphs with many valid edges, the improvement is dramatic. In the above examples we see a 7.5x speed up in a graph with many starting positions.

**Comment by shaun-edwards** [comment]
Two questions:
1. Did the graph solve get faster or was it simply the search for the lowest cost path at the end?
2. Where in your code to you identify the "true" starting node (i.e. the one after the virtual node).

File: descartes_planner/src/planning_graph.cpp [hunk 1]
```diff
@@ -630,45 +640,48 @@ bool PlanningGraph::getShortestPath(double &cost, std::list<JointTrajectoryPt>&
...
-      // if the weight of this path of less than a previous one, replace the return path
-      if (weight < cost)
+      cost = weight;
+      path.clear();
+      // Add the destination point.
+      auto current = *end;
+      // path.push_front(joint_solutions_map_[vertex_index_map[current]]);
+      // Starting from the destination point step through the predecessor map
+      // until the source point is reached.
+      while (current != virtual_vertex)
```
Comment (Jmeyer1292) [inline descartes_planner/src/planning_graph.cpp:679]: This is the spot that identifies the last real point. Essentially I follow the predecessor map back until i see the magic vertex.

Comment (Jmeyer1292) [inline descartes_planner/src/planning_graph.cpp:679]: This PR only improves the search component. It does not affect how long it takes to build the graph. The times I mentioned above are the sum of the build and search. The other pull request i added helps with this. Both changes combined brings the solution time on the last graph to 3.5 seconds.

Comment (Jmeyer1292) [inline descartes_planner/src/planning_graph.cpp:679]: @shaun-edwards @jrgnicho Is this okay to merge? Are there outstanding issues?

File: descartes_planner/src/planning_graph.cpp [hunk 2]
```diff
@@ -630,45 +640,48 @@ bool PlanningGraph::getShortestPath(double &cost, std::list<JointTrajectoryPt>&
...
-      // actual weight(cost) from start_id to end_id
-      double weight = weights[*end];
-
-      // if the weight of this path of less than a previous one, replace the return path
-      if (weight < cost)
+      cost = weight;
+      path.clear();
+      // Add the destination point.
+      auto current = *end;
+      // path.push_front(joint_solutions_map_[vertex_index_map[current]]);
```
Comment (jrgnicho) [inline descartes_planner/src/planning_graph.cpp]: Is this a commented out line or was it left here for documentation purposes?

File: descartes_planner/src/planning_graph.cpp [hunk 3]
```diff
@@ -630,45 +640,48 @@ bool PlanningGraph::getShortestPath(double &cost, std::list<JointTrajectoryPt>&
...
+      weight_map(get(&JointEdge::transition_cost, dg_)).distance_map(
+        boost::make_iterator_property_map(weights.begin(), get(boost::vertex_index, dg_))).predecessor_map(
+        &predecessors[0]));
 
-    for (auto end = end_points.begin(); end != end_points.end(); ++end)
+  // Search the ending points for a minimum point
+  for (auto end = end_points.begin(); end != end_points.end(); ++end)
+  {
+    double weight = weights[*end];
+    // if the weight of this path of less than a previous one, replace the return path
```
Comment (jrgnicho) [inline descartes_planner/src/planning_graph.cpp]: Should this line be
"if the weight of this path _**is**_ less ..."

File: descartes_planner/src/planning_graph.cpp [hunk 4]
```diff
@@ -630,45 +640,48 @@ bool PlanningGraph::getShortestPath(double &cost, std::list<JointTrajectoryPt>&
...
+      virtual_vertex, // start from our fake point
+      weight_map(get(&JointEdge::transition_cost, dg_)).distance_map(
+        boost::make_iterator_property_map(weights.begin(), get(boost::vertex_index, dg_))).predecessor_map(
+        &predecessors[0]));
 
-    for (auto end = end_points.begin(); end != end_points.end(); ++end)
+  // Search the ending points for a minimum point
+  for (auto end = end_points.begin(); end != end_points.end(); ++end)
+  {
+    double weight = weights[*end];
```
Comment (jrgnicho) [inline descartes_planner/src/planning_graph.cpp:669]: Out of curiosity, is this weight value the cost for the entire trajectory from the virtual point to this end point or just the cost of the last segment of this candidate trajectory?

Comment (Jmeyer1292) [inline descartes_planner/src/planning_graph.cpp:669]: It's the former: the cost from the virtual start position to this final vertex.

File: descartes_planner/src/planning_graph.cpp (hunk 2 above)
Comment (Jmeyer1292) [inline descartes_planner/src/planning_graph.cpp]: An oversight, I'll clean it up. +1.

File: descartes_planner/src/planning_graph.cpp (hunk 4 above)
Comment (jrgnicho) [inline descartes_planner/src/planning_graph.cpp:669]: This is a bit confusing to me as the **weight** array is initialized to a size equal to **num_vert**, which I think is the number of trajectory points.  Shouldn't this array's size be set to the number of trajectories or end points?  Functionally, this probably doesn't impact the final result.

**Comment by Jmeyer1292** [comment]
@jrgnicho I have corrected the commented lines you pointed out and re-based.

**Comment by jrgnicho** [comment]
you can merge

**Comment by Jmeyer1292** [comment]
Thank you for reviewing.

**Pull request merged**